src/main/java/uk/nhs/digital/nhsconnect/lab/results/translator/mapper/DiagnosticReportMapper.java

```diff
@@ -1,12 +1,133 @@
+        private final Patient patient;
+        private final List<Specimen> specimens;
+        private final List<Observation> observations;
+        private final Practitioner performingPractitioner;
+        private final Organization performingOrganization;
+        private final ProcedureRequest procedureRequest;
+        private DiagnosticReport fhir;
+        private ServiceReportDetails serviceReportDetails;
+
+        public DiagnosticReport map() {
+            this.fhir = new DiagnosticReport();
+            this.serviceReportDetails = message.getServiceReportDetails();
+
+            fhir.setId(uuidGenerator.generateUUID());
+            // fhir.issued = SG2.DTM
+            mapIssued();
+            // fhir.basedOn
+            mapBasedOn();
+            // fhir.status
+            fhir.setStatus(STATUS_MAP.get(serviceReportDetails.getStatus().getEvent()));
+            // fhir.identifier
+            mapIdentifier();
+            // fhir.code
+            mapCode();
+            // fhir.subject
+            fhir.getSubject().setReference(fullUrlGenerator.generate(patient));
+            // fhir.specimens
+            mapSpecimens();
+            // fhir.results
+            mapObservations();
+            // fhir.performer
+            mapPerformer();
+            /* TODO: Add BasedOn - ProcedureReport & Result - Observation */
+
+            return fhir;
+        }
+
+        private void mapBasedOn() {
+            Optional.ofNullable(procedureRequest).ifPresent(request ->
+                fhir.addBasedOn().setReference(fullUrlGenerator.generate(request)));
+        }
+
+        private void mapIssued() {
+            LocalDateTime dateIssued = serviceReportDetails.getDateIssued().getDateIssued();
+            ZonedDateTime zonedDateTime = dateIssued.atZone(TimestampService.UK_ZONE);
+            fhir.setIssued(Date.from(zonedDateTime.toInstant()));
+        }
+
+        private void mapIdentifier() {
+            final Identifier identifier = new Identifier();
+            identifier.setValue(serviceReportDetails.getReference().getNumber());
+            fhir.addIdentifier(identifier);
+        }
+
+        private void mapSpecimens() {
+            specimens.stream()
+                .map(fullUrlGenerator::generate)
+                .forEach(reference -> fhir.addSpecimen().setReference(reference));
+        }
+
+        private void mapCode() {
+            fhir.getCode().addCoding()
+                .setDisplay(CODE_DISPLAY)
+                .setCode(CODE_NUMBER)
+                .setSystem(CODE_SYSTEM);
+        }
+
+        private void mapPerformer() {
+            Optional.ofNullable(performingOrganization).ifPresent(organization ->
+                fhir.addPerformer().getActor().setReference(fullUrlGenerator.generate(organization)));
+            Optional.ofNullable(performingPractitioner).ifPresent(performer ->
+                fhir.addPerformer().getActor().setReference(fullUrlGenerator.generate(performer)));
```

Comment:
For the sake of only writing the functional code once, try a stream
```suggestion
            Stream.of(performingOrganization, performingPractitioner)
                .filter(Objects::nonNull)
                .map(fullUrlGenerator::generate)
                .forEach(performerUrl -> fhir.addPerformer().getActor().setReference(performerUrl));
```

Reply:
Done in acfd0b8